### ISSUE TITLE
Add captive portal support via dhcp option 114 (RFC 8910, systemd-networkd)

### ISSUE THREAD

**Issue by kmille** [systemd/systemd#25517]
### Component

systemd-networkd

### Is your feature request related to a problem? Please describe

Hey,

I'm currently sitting in the German Flixtrain and have trouble getting internet access using the train's wifi. The problem is that systemd-networkd currently does not support RFC 8910 Captive-Portal Identification in DHCP and Router Advertisements (RAs) [0][1].

Checking the current main branch (fe5779cf3f939d8e1ed21ab621dddde0b6699874), it does not seem like it is implemented by systemd-network (option 114):

```kmille@linbox:systemd rg SD_DHCP_OPTION_STATIC_ROUTE
src/libsystemd-network/sd-dhcp-client.c
153:        SD_DHCP_OPTION_STATIC_ROUTE,                    /* 33 */

src/libsystemd-network/test-dhcp-client.c
86:        /* RFC7844: option 33 (SD_DHCP_OPTION_STATIC_ROUTE) is set in the

src/libsystemd-network/sd-dhcp-lease.c
704:        case SD_DHCP_OPTION_STATIC_ROUTE:

src/systemd/sd-dhcp-client.h
83:        SD_DHCP_OPTION_STATIC_ROUTE                   = 33,  /* [RFC2132] */

src/network/networkd-dhcp4.c
1397:                        r = sd_dhcp_client_set_request_option(link->dhcp_client, SD_DHCP_OPTION_STATIC_ROUTE);
kmille@linbox:systemd rg SD_DHCP_OPTION_DHCP_CAPTIVE_PORTAL
src/systemd/sd-dhcp-client.h
158:        SD_DHCP_OPTION_DHCP_CAPTIVE_PORTAL            = 114, /* [RFC8910] */
kmille@linbox:systemd
```
(`SD_DHCP_OPTION_STATIC_ROUTE` is implemented, `SD_DHCP_OPTION_DHCP_CAPTIVE_PORTAL` not.)

[0] https://datatracker.ietf.org/doc/rfc8910/
[1] https://developer.android.com/about/versions/11/features/captive-portal

\### NOT SUPER RELEVANT HERE ###

The odd thing in my case is the redirection. Firefox detects a captive portal, but the redirected page does not show a form to accept the terms of services (only showed visiting https://media.flixtrain.com/landing.html).

```GET /canonical.html HTTP/1.1
Host: detectportal.firefox.com
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:107.0) Gecko/20100101 Firefox/107.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Cache-Control: no-cache
Pragma: no-cache
Connection: keep-alive

HTTP/1.1 302 Found
Server: squid
Date: Thu, 24 Nov 2022 11:51:09 GMT
Content-Length: 0
Location: https://media.flixtrain.com/redirect?url=http%3A%2F%2Fdetectportal.firefox.com%2Fcanonical.html
X-Cache: MISS from IMP-redirect
X-Cache-Lookup: MISS from IMP-redirect:3129
Via: 1.1 IMP-redirect (squid)
Connection: close
```

But it supports `RFC 8910`: 

![2022-11-24-134617_screenshot](https://user-images.githubusercontent.com/6470600/203800115-631d78de-9ac1-4fc6-beee-c3b4d94b476d.png)

```bash
kmille@linbox:~ curl https://media.flixtrain.com/services/cna/capport
{"captive":false,"user-portal-url":"https://media.flixtrain.com/services/cna/landing?type=capport","venue-info-url":"https://media.flixtrain.com/"}%
```

`https://media.flixtrain.com/services/cna/landing?type=capport"` redirects to `https://media.flixtrain.com/landing.html`.

### Describe the solution you'd like

Do you implement this? What do you think is a good idea to handle this?
We can show the captive portal url by the `networkctl` tool. Do you think running a `xdg-open <url>` is a valid solution?

**Comment by TriMoon** [comment]
> Do you think running a `xdg-open <url>` is a valid solution?

That is a "No-NOO" because not all installed systems have a GUI by default...
But even when the current system does, it still is bad practive to automatically open an URL in the users browser.

The max that could be allowable, IMHO, is a notification telling the user that a captive portal was detected and it needs manual interaction.
Keep in mind that systemd is completely unware of any GUI the user might run...

**Comment by 21hertz** [comment]
Yes, detection via RFC8910 is needed for many reasons. No GUI is needed to support RFC8908 and RFC8910, but some measures _should_ be taken by any network client subsystem that handles networking/network connections if the DHCP options are received from the networks DHCP-servers.

Just to be clear, I have very little knowledge about the inner workings of systemd or how other subsystems
integrates with it. But I have been working with server/network-side for guest networks
including captive portals for 15+ years. I always had the goal to support as many client OS'es as technically possible,
including the about 4-5% of users running Linux-based distributions.
And to make the onboarding process for end-users to be reliable, user-friendly and secure.


RFC8908's API data structure is simple with only two required keys presented in JSON:

1. **captive** (boolean). Indicates whether the client is in a state of captivity, i.e, it has not satisfied the conditions to access the external network. If the client is captive (i.e., captive=true), it will still be allowed enough access for it to perform server authentication.

2. **user-portal-url** (string). Provides the URL of a web portal that MUST be accessed over TLS. With which a user can interact.

In other words, these RFC's will get the captiveportal enduser a better experience because it's: 
- Fast. The client gets the URL to the Captiveportal API (JSON) directly on the first packet respond from DHCP-server the same second as the it gets it's IP-address etc). Client OS can react immediately (of course optional).
- Reliable. No workarounds needed for clients to find out what the captiveportal URL is. No MITM attacks needed. No DNS issues. 
- Secure. HTTPS. The Captiveportal API is required to be using TLS (FQDN with valid certificate). Also, the captiveportal website itself is required to be accessed over TLS. 
- Already adapted by Apple and Android. Majority of organizations using captive portals also already implemented support.


Anyways. Client OS should handle the response in _some way_. Some options I can see:

1. Report to user/system what the received DHCP options has been set to. Let the user decide what to do.
Maybe just inform client system + user via already established methods of informing users of important network events or information about the network they connected to (Warning, Notice or Informational level?)

2. Report to system what the received DHCP options has been set to and send a http request to the specified Captiveportal API and check if captive is set to true or false. If true, do something with the result, something GUI/browser agnostic. Inform user/system that it is behind a captive portal and what the URL for the captiveportal is. User can still use a text-based browser (or whatever method) to visit the captive portal page (if captiveportal page designer is stupid enough to have javascript or other non-compatible is totally another issue).

3. The above + a user that connects to a network and is running a GUI, should be interactively be informed via another service that knows the GUI or default browser. That other service could actively check if network/systemd has reported that captiveportal API has set captive=true.  If true, ask user if the user-portal-url should be opened using the default open browser. Would not even have to be auto-opened browser like it is in other OS'es. Just a interactive message box that let users click Yes or No when asked if they want to open the URL in their default browser (or let user choose browser/app to open the URL with).

...just my 2 cents.

If anyone got any more info or news about implementation/support for RFC8910 and RFC8908 in systemd (or other services used by Linux in large), I'd be happy to know about it.

Quote RFC8910:
_"By removing or reducing the need for captive portals to perform MITM hijacking, this mechanism improves security by making the portal and its actions visible, rather than hidden, and reduces the likelihood that users will disable useful security safeguards like DNSSEC validation, VPNs, etc. in order to interact with the captive portal."_

**Comment by mervn** [comment]
Any interest in this? The comment from @21hertz was insightful.

If I'm understanding this correctly, there is at least one option:

- **Integrate `signals: CaptiveChanged(b captive, s user-portal-url)` into systemd-networkd**. networkd itself will act as a RFC8910 client and forward the RFC8908 compliant data, through dbus. This comes with the upside of being able to "enable" captive detection per network.

> This can also be put into a separate service, named systemd-captived.

This uses existing infrastructure (dbus), does not mandate a GUI, and allows for multiple listeners to react simultaneously; this last point, in particular, means that users no longer have to choose between a notification or a response - they can have both.
